net/dns/resolver/forwarder_test.go

```diff
@@ -240,3 +251,221 @@ func FuzzClampEDNSSize(f *testing.F) {
 		clampEDNSSize(data, maxResponseBytes)
 	})
 }
+
+func runDNSServer(tb testing.TB, response []byte, onRequest func(bool, []byte)) (port uint16) {
+	tcpResponse := make([]byte, len(response)+2)
+	binary.BigEndian.PutUint16(tcpResponse, uint16(len(response)))
+	copy(tcpResponse[2:], response)
+
+	// Repeatedly listen until we can get the same port.
+	const tries = 25
+	var (
+		tcpLn *net.TCPListener
+		udpLn *net.UDPConn
+		err   error
+	)
+	for try := 0; try < tries; try++ {
+		if tcpLn != nil {
+			tcpLn.Close()
+			tcpLn = nil
+		}
+
+		tcpLn, err = net.ListenTCP("tcp4", &net.TCPAddr{
+			IP:   net.IPv4(127, 0, 0, 1),
+			Port: 0, // Choose one
+		})
+		if err != nil {
+			tb.Fatal(err)
+		}
+		udpLn, err = net.ListenUDP("udp4", &net.UDPAddr{
+			IP:   net.IPv4(127, 0, 0, 1),
+			Port: tcpLn.Addr().(*net.TCPAddr).Port,
+		})
+		if err == nil {
+			break
+		}
+	}
+	if tcpLn == nil || udpLn == nil {
+		if tcpLn != nil {
+			tcpLn.Close()
+		}
+		if udpLn != nil {
+			udpLn.Close()
+		}
+
+		// Skip instead of being fatal to avoid flaking on extremely
+		// heavily-loaded CI systems.
+		tb.Skipf("failed to listen on same port for TCP/UDP after %d tries", tries)
+	}
+
+	port = uint16(tcpLn.Addr().(*net.TCPAddr).Port)
+
+	handleConn := func(conn net.Conn) {
+		defer conn.Close()
+
+		// Read the length header, then the buffer
+		var length uint16
+		if err := binary.Read(conn, binary.BigEndian, &length); err != nil {
+			tb.Logf("error reading length header: %v", err)
+			return
+		}
+		req := make([]byte, length)
+		n, err := io.ReadFull(conn, req)
+		if err != nil {
+			tb.Logf("error reading query: %v", err)
+			return
+		}
+		req = req[:n]
+		onRequest(true, req)
+
+		// Write response
+		if _, err := conn.Write(tcpResponse); err != nil {
+			tb.Logf("error writing response: %v", err)
+			return
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			conn, err := tcpLn.Accept()
+			if err != nil {
+				return
+			}
+			go handleConn(conn)
+		}
+	}()
+
+	handleUDP := func(addr netip.AddrPort, req []byte) {
+		onRequest(false, req)
+		if _, err := udpLn.WriteToUDPAddrPort(response, addr); err != nil {
+			tb.Logf("error writing response: %v", err)
+		}
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			buf := make([]byte, 65535)
+			n, addr, err := udpLn.ReadFromUDPAddrPort(buf)
+			if err != nil {
+				return
+			}
+			buf = buf[:n]
+			go handleUDP(addr, buf)
+		}
+	}()
+
+	tb.Cleanup(func() {
+		tcpLn.Close()
+		udpLn.Close()
+		tb.Logf("waiting for listeners to finish...")
+		wg.Wait()
+	})
+	return
+}
+
+func TestForwarderTCPFallback(t *testing.T) {
+	envknob.Setenv("TS_DEBUG_DNS_FORWARD_SEND", "true")
```

Comment:
unset after test?

Reply:
Good catch; had it read and restore the prior value after the test 👍